src/Adapter/Product/ProductCustomizationFieldUpdater.php

```diff
@@ -0,0 +1,154 @@
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product;
+
+use CustomizationField;
+use PrestaShop\PrestaShop\Core\Domain\Product\Customization\CustomizationFieldDeleterInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Customization\ValueObject\CustomizationFieldId;
+use PrestaShop\PrestaShop\Core\Domain\Product\Customization\ValueObject\CustomizationFieldType;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\CannotUpdateProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ProductCustomizabilitySettings;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use Product;
+
+/**
+ * Updates CustomizationField & Product relation
+ */
+class ProductCustomizationFieldUpdater
+{
+    /**
+     * @var CustomizationFieldPersister
+     */
+    private $customizationFieldPersister;
+
+    /**
+     * @var CustomizationFieldDeleterInterface
+     */
+    private $customizationFieldDeleter;
+
+    /**
+     * @var ProductProvider
+     */
+    private $productProvider;
+
+    /**
+     * @var ProductPersister
+     */
+    private $productPersister;
+
+    /**
+     * @param CustomizationFieldPersister $customizationFieldPersister
+     * @param CustomizationFieldDeleterInterface $customizationFieldDeleter
+     * @param ProductProvider $productProvider
+     * @param ProductPersister $productPersister
+     */
+    public function __construct(
+        CustomizationFieldPersister $customizationFieldPersister,
+        CustomizationFieldDeleterInterface $customizationFieldDeleter,
+        ProductProvider $productProvider,
+        ProductPersister $productPersister
+    ) {
+        $this->customizationFieldPersister = $customizationFieldPersister;
+        $this->customizationFieldDeleter = $customizationFieldDeleter;
+        $this->productPersister = $productPersister;
+        $this->productProvider = $productProvider;
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param CustomizationField[] $customizationFields
+     */
+    public function setProductCustomizationFields(ProductId $productId, array $customizationFields): void
```

Comment:
I think you could `Product $product` as the parameter here, since you already fetch the product in the handler, and it would remove on dependency with `ProductProvider` in this service

I guess using ValueObject IDs parameters are useful when dealing with "pure" relationship that we can manage with inserting directly in the table

Reply:
So now that the Updater is in charge of update the product it makes sense that it has the `productRepository` since it needs it in `refreshProductCustomizability`

But now it's the opposite, the `SetProductCustomizationFieldsHandler` maybe doesn't need it any more, because it doesn't use the Product instance as an input nor anywhere else it only uses it to create the array it returns.

Should this array_map be moved into this Updater class? Or does it make more sense for the Handler to manage itself the format of its return?